File: 8vim/src/main/java/inc/flide/vim8/structures/yaml/Layers.java
package inc.flide.vim8.structures.yaml;

import com.fasterxml.jackson.annotation.JsonProperty;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class Layers {
    public List<Action> hidden = new ArrayList<>();
    @JsonProperty("default")
    public Layer defaultLayer;
    public Map<ExtraLayer, Layer> extraLayers = new HashMap<>();
}

File: 8vim/src/main/kotlin/inc/flide/vim8/ime/layout/models/yaml/versions/version2/Layers.kt
package inc.flide.vim8.ime.layout.models.yaml.versions.version2

import arrow.core.None
import arrow.core.Option
import arrow.optics.optics
import com.fasterxml.jackson.annotation.JsonProperty
import inc.flide.vim8.ime.layout.models.yaml.versions.common.Action
import inc.flide.vim8.ime.layout.models.yaml.versions.common.ExtraLayer
import inc.flide.vim8.ime.layout.models.yaml.versions.common.Layer
import inc.flide.vim8.ime.layout.models.yaml.versions.version21.Layers as Layers21
import java.util.EnumMap

@optics
data class Layers(
    val hidden: List<Action> = ArrayList(),
    @JsonProperty("default") val defaultLayer: Option<Layer> = None,
    val extraLayers: Map<ExtraLayer, Layer> = EnumMap(ExtraLayer::class.java)
) {
    companion object
}

fun Layers.toLatest(): Layers21 = Layers21(
    hidden = hidden,
    defaultLayer = defaultLayer,
    extraLayers = extraLayers
)
